Develop > Toolkit > Blockchain > Spawn Testing Chain

dawnkelly09 opened the pull request:
This PR:
-uses Zombienet material from PR #31 to create this page for the new IA
- moves snippets & images, updates paths
- moved code out to snippets
- some formatting changes

themacexpert commented on develop/toolkit/blockchain/spawn-testing-chain.md:
````diff
@@ -1,4 +1,1213 @@
+
+    First, clone the OpenZeppelin template repository:
+
+    ```bash
````
Missing a closing ``` here on this command

themacexpert commented on develop/toolkit/blockchain/spawn-testing-chain.md:
```diff
@@ -1,4 +1,1213 @@
+- `enable_tracing?` ++"boolean"++ - enable the tracing system. Only available on Kubernetes. Defaults to `true`
+- `provider` ++"string"++ - provider to use. Default is `kubernetes`"
+- `polkadot_introspector?` ++"boolean"++ - deploy an instance of polkadot-introspector. Only available on Podman and Kubernetes. Defaults to `false`
+- `backchannel?` ++"boolean"++ - deploy an instance of backchannel server. Only available on Kubernetes. Defaults to `false`
```
I'll let @eshaben weigh in on this ... Do we want to include the ? to show optionality or better to explicitly write out, `Optional`

eshaben replied:
Yeah, explicitly write it out:

```
- `backchannel` ++"boolean"++ - *(optional)* deploy an instance of backchannel server. Only available on Kubernetes. Defaults to `false`
```

themacexpert commented on .snippets/code/develop/toolkit/blockchain/spawn-testing-chain/relaychain-example-nodes.toml:
General formatting question, when we have a path like this: 
chain_spec_path = "/path/to/chain-spec.json

Do we want to say that or INSERT_PATH_TO_CHAIN_SPEC.JSON ?

dawnkelly09 replied:
I changed it. I like your version better and it will look more consistent with other variables used

themacexpert commented on develop/toolkit/blockchain/spawn-testing-chain.md:
````diff
@@ -1,4 +1,1213 @@
+    zombienet version
+    ```
+
+    So then, you can refer to the `zombienet` executable directly:
````
```suggestion
```

themacexpert commented on develop/toolkit/blockchain/spawn-testing-chain.md:
````diff
@@ -1,4 +1,1213 @@
+    So then, you can refer to the `zombienet` executable directly:
+
+    ```bash
+    zombienet version
````
```suggestion
```

themacexpert commented on develop/toolkit/blockchain/spawn-testing-chain.md:
````diff
@@ -1,4 +1,1213 @@
+
+    So then, you can refer to the `zombienet` executable directly:
+
+    ```bash
````
```suggestion
```

themacexpert commented on develop/toolkit/blockchain/spawn-testing-chain.md:
````diff
@@ -1,4 +1,1213 @@
+
+    ```bash
+    zombienet version
+    ```
````
```suggestion
```

themacexpert commented on develop/toolkit/blockchain/spawn-testing-chain.md:
```diff
@@ -1,4 +1,1213 @@
+        - `<networkConfig>` - a file that declares the desired network to be spawned in `.toml` or `.json` format. For further information, check out the [Configuration Files](#configuration-files) section
+
+    !!! warning
+        For the `spawn` command to work on macOS, users need to be aware that the Polkadot binary is currently not compatible with macOS. As a result, macOS users will need to clone the [Polkadot repository](https://github.com/paritytech/polkadot-sdk){target=_blank}, build Polkadot binary, and manually add it to their PATH.
```
```suggestion
        For the `spawn` command to work on macOS, users need to be aware that the Polkadot binary is currently not compatible with macOS. As a result, macOS users will need to clone the [Polkadot repository](https://github.com/paritytech/polkadot-sdk){target=_blank}, build the Polkadot binary, and manually add it to their PATH.
```

themacexpert commented on develop/toolkit/blockchain/spawn-testing-chain.md:
```diff
@@ -1,4 +1,1213 @@
+
+Zombienet releases are available on the [Zombienet repository](https://github.com/paritytech/zombienet){target=_blank}.
+
+To install Zombienet, there are multiple options available, depending on the user's preferences and the environment where it will be used. The following section will guide you through the installation process for each of the available options.
```
Can you double check this line?

For some reason in my .md file of this, it shows as 
`Toinstall Zombienet, there are multiple options available, depending on the user's preferences and the environment where it will be used. The following section will guide you through the installation process for each of the available options.`

 but correctly renders on Github. Not sure why, if you don't have the same don't worry

dawnkelly09 replied:
Looks fine when I spin it up.

themacexpert reviewed (changes requested):
Made a quick initial pass, will try to return to this and see if I see anything else

eshaben reviewed (commented):
I think this page needs to be split up. I looked back at the old PR, I think we should keep the pages split up as they were.

Under Develop:
- Get Started
- Write Tests

Under Tutorials:
- Spawn a Basic Network

eshaben commented on develop/toolkit/blockchain/.pages:
```diff
@@ -1,5 +1,5 @@
   - index.md
-  # - 'Spawn Chains for Testing with Zombienet': spawn-testing-chain.md
-  - 'Fork Live Chains with Chopsticks': fork-live-chains.md
+  - spawn-networks
+  - fork-chains
```
I feel like we should stick to chains or networks for these.

eshaben commented on develop/toolkit/blockchain/fork-chains/chopsticks/.pages:
```diff
@@ -0,0 +1,4 @@
+title: Chopsticks
+nav:
+  - index.md
+  - 'Fork Live Chains with Chopsticks': fork-live-chains.md
```
This ideally should be a Get Started page.

If we take a step back and look at the structure between Chopsticks and Zombienet, Chopsticks includes install, configuration, and examples. Whereas now in the Zombienet section, there is a page for Installation and a page for Configuration. We need to be consistent in how we document tools, which is why I proposed the structure I did in this [comment](https://github.com/polkadot-developers/polkadot-docs/pull/154#pullrequestreview-2443955760)

If I made some contradicting statement somewhere along the line, I apologize

eshaben commented on develop/toolkit/blockchain/spawn-networks/index.md:
```diff
@@ -0,0 +1,7 @@
+---
+title: Spawn Networkds for Testing
```
```suggestion
title: Spawn Networks for Testing
```

eshaben commented on develop/toolkit/blockchain/spawn-networks/zombienet/spawn-basic-network.md:
```diff
@@ -0,0 +1,312 @@
+
+To successfully complete this tutorial, you must ensure you've first:
+
+- [Installed Zombienet](/develop/toolkit/blockchain/spawn-networks/zombienet/installation.md){target=\_blank}
+- Reviewed the information in [Configuration](/develop/toolkit/blockchain/spawn-networks/zombienet/configuration.md){target=\_blank} and understand how to customize a Zombienet spawned network
```
Please check all links to make sure they don't end in `.md`

eshaben commented on develop/toolkit/blockchain/spawn-networks/zombienet/spawn-basic-network.md:
```diff
@@ -0,0 +1,312 @@
+
+After running this command, you will see the logs of the `alice` node in real-time, which can be useful for debugging purposes. The logs of the `bob` and `collator01` nodes can be checked similarly.
+
+## Testing DSL
+
+Zombienet provides a Domain Specific Language (DSL) for writing tests. The DSL is designed to be human-readable and allows you to write tests using natural language expressions. You can define assertions and tests against the spawned network using this DSL. This way, users can evaluate different metrics, such as:
+
+- **On-chain storage** - the storage of each of the chains running via Zombienet
+- **Metrics** - the metrics provided by the nodes
+- **Histograms** - visual representations of metrics data
+- **Logs** - detailed records of system activities and events
+- **System events** - notifications of significant occurrences within the network
+- **Tracing** - detailed analysis of execution paths and operations
+- **Custom API calls (through Polkadot.js)** - personalized interfaces for interacting with the network
+- **Commands** - instructions or directives executed by the network
+
+These abstractions are expressed by sentences defined in a natural language style. Therefore, each test line will be mapped to a test to run. Also, the test file (`*.zndsl`) includes pre-defined header fields used to define information about the suite, such as network configuration and credentials location.
+
+## The Test File
+
+The test file is a text file with the extension `.zndsl`. It is divided into two parts: the header and the body. The header contains the network configuration and the credentials to use, while the body contains the tests to run.
+
+The header is defined by the following fields:
+
+- **`description`** ++"string"++ - long description of the test suite (optional)
+- **`network`** ++"string"++ - path to the network definition file, supported in both `.json` and `.toml` formats
+- **`creds`** ++"string"++ - credentials filename or path to use (available only with Kubernetes provider). Looks in the current directory or `$HOME/.kube/` if a filename is passed
+
+The body contains the tests to run. Each test is defined by a sentence in the DSL, which is mapped to a test to run. Each test line defines an assertion or a command to be executed against the spawned network.
```
This basic information about testing should remain it's own page and not be nested into a tutorial. I'd prefer to see the content split up as we had it before in the ecosystem docs draft https://papermoonio.github.io/polkadot-ecosystem-docs-draft/dev-tools/zombienet/testing/
https://papermoonio.github.io/polkadot-ecosystem-docs-draft/dev-tools/zombienet/tutorials/spawn-a-basic-network/

eshaben commented on develop/toolkit/blockchain/spawn-networks/zombienet/spawn-basic-network.md:
```diff
@@ -0,0 +1,312 @@
+---
+title: Spawn Testing Networks with Zombienet
+description: Learn how to use Zombienet to spawn Polkadot SDK-based blockchain networks for testing.
```
Too short

dawnkelly09 commented:
Did some work on this PR to bring it inline with the review feedback and add back the missing Chopsticks pages. This PR now does the following:

1. Creates the following pages for Zombienet:
- /develop/toolkit/blockchain/spawn-chains/zombienet/get-started.md 
- /develop/toolkit/blockchain/spawn-chains/zombienet/write-tests.md
- /tutorials/blockchains/testing/spawn-basic-chain.md
- 
2. Creates the following pages for Chopsticks:
- /develop/toolkit/blockchain/fork-chains/chopsticks/get-started.md
- /tutorials/blockchains/testing/fork-live-chains.md

3. Addresses the feedback from Kevin and Erin up to this point. 
4. Moves images and code snippets and updates paths to match new pages
5. Audits all links, all work

nhussein11 reviewed (approved):
Reviewed each file here, the look and feel is great! Thanks all @dawnkelly09 @eshaben @themacexpert !!!

LGTM! 🚀 🚀 🚀 🚀

themacexpert commented on tutorials/blockchains/testing/spawn-basic-chain.md:
````diff
@@ -0,0 +1,91 @@
+To check node logs, you can use the command listed in the output as `Log Cmd`. For instance, to check the logs of the `alice` node, you can open a new terminal and run the following command:
+
+```bash
+tail -f /var/folders/f4/7rdt2m9d7j361dm453cpggbm0000gn/T/zombie-75a01b93c92d571f6198a67bcb380fcd_21724-SEzfCidQ1za4/alice.log
````
This command is slightly different from the one you show in terminyl. I think this one here is the correct one because it ends in .log but I'll leave it up to you to confirm. 

This is the command shown in the terminyl snippet: 
tail -f /var/folders/f4/7rdt2m9d7j361dm453cpggbm0000gn/T/zombie-75a01b93c92d571f6198a67bcb380fcd_21724-2

themacexpert reviewed (approved):
1 very minor comment otherwise looks great!

eshaben reviewed (approved):
🚀
